nifi-extension-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/JWTBearerOAuth2AccessTokenProvider.java

```diff
@@ -0,0 +1,664 @@
+                            final MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
+                            final byte[] bytes = messageDigest.digest(certificates[0].getEncoded());
+                            String url = Base64.getEncoder().encodeToString(bytes);
+                            headerBuilder.x509CertSHA256Thumbprint(new Base64URL(url));
+                        }
+                    }
+                }
+            } catch (final AccessTokenRetrievalException e) {
+                throw e;
+            } catch (final Exception e) {
+                throw new AccessTokenRetrievalException("Failed to set x5t header", e);
+            }
+        }
+
+        Map<String, String> formParams = new HashMap<>();
+        formParams.put("grant_type", grantType);
+        formParams.put(assertion, getAssertion(headerBuilder.build(), claimsSetBuilder.build()));
+        formParams.putAll(this.formParams);
+
+        requestTokenEndpoint(formParams);
+    }
+
+    protected void requestTokenEndpoint(Map<String, String> formParams) throws URISyntaxException {
+        HttpRequestHeadersSpec request = webClientService
+                .post()
+                .uri(new URI(tokenEndpoint))
+                .header("Accept", APPLICATION_JSON)
+                .header("Content-Type", APPLICATION_URLENCODED)
+                .body(formParams.entrySet()
+                        .stream()
+                        .map(param -> param.getKey() + "=" + param.getValue())
+                        .collect(Collectors.joining("&")));
+
+        try (final HttpResponseEntity response = request.retrieve()) {
+            if (response.statusCode() != 200) {
+                final String message = "Failed to retrieve Access Token: HTTP %s with Response [%s]".formatted(
+                        response.statusCode(),
+                        readBodyAsText(response).orElse("[no body]"));
+
+                throw new AccessTokenRetrievalException(message);
+            }
+
+            try (final InputStream body = response.body()) {
+                accessDetails = ACCESS_DETAILS_MAPPER.readValue(body, AccessToken.class);
+            }
+        } catch (final IOException e) {
+            throw new AccessTokenRetrievalException("Failed to retrieve or process access token details", e);
+        }
+    }
+
+    protected String getAssertion(JWSHeader jwsHeader, JWTClaimsSet jwtClaimsSet) throws JOSEException {
+        SignedJWT signedJWT = new SignedJWT(jwsHeader, jwtClaimsSet);
+        signedJWT.sign(signer);
+        return signedJWT.serialize();
+    }
+
+    private Optional<String> readBodyAsText(final HttpResponseEntity response) {
+        try (final InputStream is = response.body()) {
+            final String body = new String(is.readAllBytes(), StandardCharsets.UTF_8);
+            return Optional.of(body);
+        } catch (final IOException e) {
+            getLogger().warn("Failed to read response body", e);
```

Comment:
It would be helpful to include the HTTP Status Code, and the URL if available for troubleshooting this warning.

Reply:
This method is called from from error handling in `requestTokenEndpoint` where we log the HTTP Code and response. I added the URL as well. Not sure we should also add this information in this specific warn log, no?

Reply:
Good point, in this case, I recommend removing this warning. It would probably be better to throw the exception, catch it in the calling method, and throw it as the cause of the `AccessTokenRetrievalException`.

Reply:
Fair enough, pushed a commit to have a better logic there

Reply:
Thanks, the updated approach looks good.